src/config.rs

```diff
@@ -0,0 +1,152 @@
+#[serde(rename_all = "lowercase")]
+pub enum Mode {
+    Compact,
+    Light,
+    Full,
+}
+
+impl std::fmt::Display for Mode {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Mode::Compact => write!(f, "compact"),
+            Mode::Light => write!(f, "light"),
+            Mode::Full => write!(f, "full"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, ValueEnum, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+pub enum Network {
+    Mainnet,
+    Preprod,
+    Preview,
+}
+
+#[derive(Debug, Clone, ValueEnum, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+pub enum LogLevel {
+    Debug,
+    Info,
+    Warn,
+    Error,
+    Trace,
+}
+
+// Implement conversion from LogLevel enum to tracing::Level
+impl From<LogLevel> for Level {
+    fn from(log_level: LogLevel) -> Self {
+        match log_level {
+            LogLevel::Debug => Level::DEBUG,
+            LogLevel::Info => Level::INFO,
+            LogLevel::Warn => Level::WARN,
+            LogLevel::Error => Level::ERROR,
+            LogLevel::Trace => Level::TRACE,
+        }
+    }
+}
+
+impl Config {
+    pub fn from_args_with_detector<F>(args: Args, detector: F) -> Result<Self, AppError>
+    where
+        F: Fn(&str) -> Result<Network, AppError>,
+    {
+        let node_socket_path = args
+            .node_socket_path
+            .ok_or(AppError::Server("--node-socket-path must be set".into()))?;
+
+        let icebreakers_config = if !args.solitary {
+            let reward_address = args
+                .reward_address
+                .ok_or(AppError::Server("--reward-address must be set".into()))?;
+
+            let secret = args
+                .secret
+                .ok_or(AppError::Server("--secret must be set".into()))?;
+
+            Some(IcebreakersConfig {
+                reward_address,
+                secret,
+            })
+        } else {
+            if args.reward_address.is_some() || args.secret.is_some() {
+                return Err(AppError::Server(
+                    "Cannot set --reward-address or --secret in solitary mode (--solitary)".into(),
+                ));
+            }
+            None
+        };
+
+        let network = detector(&node_socket_path)?;
+
+        Ok(Config {
+            server_address: args.server_address,
+            server_port: args.server_port,
+            log_level: args.log_level.into(),
+            node_socket_path,
+            mode: args.mode,
+            icebreakers_config,
+            max_pool_connections: 10,
+            no_metrics: args.no_metrics,
+            network,
+        })
+    }
+
+    pub fn from_args(args: Args) -> Result<Self, AppError> {
+        Self::from_args_with_detector(args, detect_network)
+    }
+}
+
+fn detect_network(socket_path: &str) -> Result<Network, AppError> {
+    let magics = genesis().all_magics();
```

Comment:
Clever! Does `mainnet` appear first? Otherwise we’d get some log noise in `cardano-node`s on Mainnet.